synctl

```diff
@@ -298,11 +317,22 @@ def main():
     action = options.action
 
     if action == "stop" or action == "restart":
-        for worker in workers:
-            stop(worker.pidfile, worker.app)
+        has_stopped = False
+
+        if len(workers) > 0:
+            stopped_worker_count = 0
+            for worker in workers:
+                if stop(worker.pidfile, worker.app):
+                    stopped_worker_count += 1
+            # If all the workers were stopped
+            if len(workers) == stopped_worker_count:
+                has_stopped = True
```

Comment:
I think the boolean logic here is a bit confusing, I'd recommend something like:

```python
has_stopped = True
for worker in workers:
    has_stopped = has_stopped and stop(worker.pidfile, worker.app)
```

And then for the main synapse we need to do something similar:

```python
if start_stop_synapse:
    has_stopped = has_stopped and stop(pidfile, "synapse.app.homeserver")
```

Reply:
Yes, do you know how I could make a worker non-responsive to test behaviour when a worker isn't stopping as expected?
I think we should still iterate over all the workers, call `stop` on each even if one isn't stopping. Leaving only the non-responsive workers up.

We can still simplify the logic to this.
```python
has_stopped = True
for worker in workers:
    if not stop(worker.pidfile, worker.app):
        # A worker doesn't stop
        has_stopped = False

if start_stop_synapse:
    if not stop(pidfile, "synapse.app.homeserver"):
        has_stopped = False
if not has_stopped:
    sys.exit(1)
```
What do you think?

Reply:
Yeah, that's pretty similar to what I had above. It should work.

I'm not sure of a way to test this reasonable.